File: app/src/main/java/com/jonasgerdes/stoppelmap/model/transportation/DepartureDay.java
package com.jonasgerdes.stoppelmap.model.transportation;

import android.support.annotation.IntDef;

import java.util.Calendar;

import io.realm.RealmList;
import io.realm.RealmObject;

/**
 * Created by Jonas on 08.07.2016.
 */
public class DepartureDay extends RealmObject {


    public static final int DAY_INVALID = -1;
    public static final int DAY_THURSDAY = 0;
    public static final int DAY_FRIDAY = 1;
    public static final int DAY_SATURDAY = 2;
    public static final int DAY_SUNDAY = 3;
    public static final int DAY_MONDAY = 4;
    public static final int DAY_TUESDAY = 5;
    //Hour next day starts (ex: 6 means if it's 4am it' still the day before)
    private static final int FIRST_HOUR_OF_NEW_DAY = 6;

    @IntDef({
            DAY_INVALID,
            DAY_THURSDAY,
            DAY_FRIDAY,
            DAY_SATURDAY,
            DAY_SUNDAY,
            DAY_MONDAY,
            DAY_TUESDAY
    })
    public @interface Day {
    }

    private
    @Day
    int day;
    private RealmList<Departure> departures;
    private String comment;

    public DepartureDay() {
    }

    public DepartureDay(@Day int day, RealmList<Departure> departures, String comment) {
        this.day = day;
        this.departures = departures;
        this.comment = comment;
    }

    public
    @Day
    int getDay() {
        return day;
    }

    public void setDay(int day) {
        this.day = day;
    }

    public RealmList<Departure> getDepartures() {
        return departures;
    }

    public void setDepartures(RealmList<Departure> departures) {
        this.departures = departures;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    @Override
    public String toString() {
        return "DepatureDay{" +
                "day=" + day +
                ", depatures=" + departures +
                ", comment='" + comment + '\'' +
                '}';
    }

    public static
    @DepartureDay.Day
    int getDayFromCalendar(Calendar date) {
        if (date.get(Calendar.MONTH) == Calendar.AUGUST
                && date.get(Calendar.DAY_OF_MONTH) >= 11
                && date.get(Calendar.DAY_OF_MONTH) <= 16) {
            int dayOfMonth = date.get(Calendar.DAY_OF_MONTH);


            if (date.get(Calendar.HOUR_OF_DAY) < FIRST_HOUR_OF_NEW_DAY) {
                dayOfMonth -= 1;
            }

            switch (dayOfMonth) {
                case 11:
                    return DAY_THURSDAY;
                case 12:
                    return DAY_FRIDAY;
                case 13:
                    return DAY_SATURDAY;
                case 14:
                    return DAY_SUNDAY;
                case 15:
                    return DAY_MONDAY;
                case 16:
                    return DAY_TUESDAY;
            }
        }
        return DAY_INVALID;
    }
}

File: app/src/main/java/com/jonasgerdes/stoppelmap/model/transportation/DepartureDay.kt
package com.jonasgerdes.stoppelmap.model.entity

import io.realm.RealmList
import io.realm.RealmObject
import java.util.*


/**
 * @author Jonas Gerdes <dev@jonasgerdes.com>
 * @since 09.08.2017
 */
open class DepartureDay : RealmObject() {
    var comment: String? = null
    var depatures: RealmList<Departure> = RealmList()

    companion object {
        val DAY_INVALID = -1
        val DAY_THURSDAY = 0
        val DAY_FRIDAY = 1
        val DAY_SATURDAY = 2
        val DAY_SUNDAY = 3
        val DAY_MONDAY = 4
        val DAY_TUESDAY = 5
        //Hour next day starts (ex: 6 means if it's 4am it' still the day before)
        private val FIRST_HOUR_OF_NEW_DAY = 6
        private val FIRST_DAY = 10
        private val LAST_DAY = 15

        fun getDayFromCalendar(date: Calendar): Int {
            if (date.get(Calendar.MONTH) == Calendar.AUGUST
                    && date.get(Calendar.DAY_OF_MONTH) >= FIRST_DAY
                    && date.get(Calendar.DAY_OF_MONTH) <= LAST_DAY) {
                var dayOfMonth = date.get(Calendar.DAY_OF_MONTH)


                if (date.get(Calendar.HOUR_OF_DAY) < FIRST_HOUR_OF_NEW_DAY) {
                    dayOfMonth -= 1
                }

                when (dayOfMonth) {
                    11 -> return DAY_THURSDAY
                    12 -> return DAY_FRIDAY
                    13 -> return DAY_SATURDAY
                    14 -> return DAY_SUNDAY
                    15 -> return DAY_MONDAY
                    16 -> return DAY_TUESDAY
                }
            }
            return DAY_INVALID
        }
    }
}